apps/emails-and-messages/src/modules/trpc/protected-client-procedure.ts

```diff
@@ -31,89 +37,62 @@ const attachAppToken = middleware(async ({ ctx, next }) => {
 
   return next({
     ctx: {
-      appToken: authData.token,
-      saleorApiUrl: authData.saleorApiUrl,
-      appId: authData.appId,
+      authData,
     },
   });
 });
 
-const validateClientToken = middleware(async ({ ctx, next, meta }) => {
-  logger.debug(
-    {
-      permissions: meta?.requiredClientPermissions,
-    },
-    "Calling validateClientToken middleware with permissions required"
-  );
-
-  if (!ctx.token) {
-    throw new TRPCError({
-      code: "INTERNAL_SERVER_ERROR",
-      message: "Missing token in request. This middleware can be used only in frontend",
+const validateClientToken = attachSaleorAuthData.unstable_pipe(
+  async ({ ctx: { authData, appBridgeToken }, next, meta }) => {
+    const logger = createLogger({
+      name: "validateClientToken",
+      requiredPermissions: meta?.requiredClientPermissions,
+      saleorApiUrl: authData.saleorApiUrl,
     });
-  }
 
-  if (!ctx.appId) {
-    throw new TRPCError({
-      code: "INTERNAL_SERVER_ERROR",
-      message: "Missing appId in request. This middleware can be used after auth is attached",
-    });
-  }
+    if (!appBridgeToken) {
+      logger.debug("Missing JWT token in the request.");
+      throw new TRPCError({
+        code: "UNAUTHORIZED",
+        message: "Missing JWT token in the request.",
+      });
+    }
 
-  if (!ctx.saleorApiUrl) {
-    throw new TRPCError({
-      code: "INTERNAL_SERVER_ERROR",
-      message:
-        "Missing saleorApiUrl in request. This middleware can be used after auth is attached",
-    });
-  }
+    logger.debug("Verify JWT token");
 
-  if (!ctx.ssr) {
     try {
-      logger.debug("trying to verify JWT token from frontend");
-      logger.debug({ token: ctx.token ? `${ctx.token[0]}...` : undefined });
-
       await verifyJWT({
-        appId: ctx.appId,
-        token: ctx.token,
-        saleorApiUrl: ctx.saleorApiUrl,
+        appId: authData.appId,
+        token: appBridgeToken,
+        saleorApiUrl: authData.saleorApiUrl,
         requiredPermissions: meta?.requiredClientPermissions ?? [],
       });
     } catch (e) {
-      logger.debug("JWT verification failed, throwing");
+      logger.debug("JWT verification failed, throwing error");
       throw new ProtectedHandlerError("JWT verification failed: ", "JWT_VERIFICATION_FAILED");
     }
-  }
 
-  return next({
-    ctx: {
-      ...ctx,
-      saleorApiUrl: ctx.saleorApiUrl,
-    },
-  });
-});
+    return next();
+  }
+);
 
 /**
- * Construct common graphQL client and attach it to the context
+ * Construct common GraphQL client and attach it to the context
  *
  * Can be used only if called from the frontend (react-query),
  * otherwise jwks validation will fail (if createCaller used)
  *
  * TODO Rethink middleware composition to enable safe server-side router calls
  */
 export const protectedClientProcedure = procedure
-  .use(attachAppToken)
+  .use(attachLogger)
```

Comment:
did you replace appToken middleware with logger? What about auth data now?

Reply:
logger is executed as first, before any auth check (as intended).

Auth data are added to the context with `validateClientToken` middleware.

Since validation depends on attaching the auth data, validation middleware use `unstable_pipe` feature to check it first. Previously code for checking tokens in context was doubled, with this change that's no longer the case and types propagate (tokens are no longer marked as optional in the context after the validation)

Reply:
Do we have to use `ubstable_pipe`? Can we achieve some declarative composition like
```
procedure
  .use(attachLogger)
  .use(attachAuthData)
  .use(authorize) // etc

```

Reply:
@krzysztofwolski I've read about unstable_pipe, I know how it works now 👍